extensions/amp-list/0.1/amp-list.js

```diff
@@ -204,60 +214,65 @@ export class AmpList extends AMP.BaseElement {
       // Remove any previous items before the reload
       removeChildren(dev().assertElement(this.container_));
     }
-
-    const itemsExpr = this.element.getAttribute('items') || 'items';
-    return this.fetch_(itemsExpr).then(items => {
-      if (this.element.hasAttribute('single-item')) {
-        user().assert(typeof items !== 'undefined' ,
-            'Response must contain an array or object at "%s". %s',
+    if (this.templateRendererHelper_.canRenderTemplates()) {
+      this.fetchAndRenderTemplate_();
+    } else {
+      const itemsExpr = this.element.getAttribute('items') || 'items';
+      return this.fetch_(itemsExpr).then(items => {
+        if (this.element.hasAttribute('single-item')) {
+          user().assert(typeof items !== 'undefined',
+              'Response must contain an array or object at "%s". %s',
+              itemsExpr, this.element);
+          if (!isArray(items)) {
+            items = [items];
+          }
+        }
+        user().assert(isArray(items),
+            'Response must contain an array at "%s". %s',
             itemsExpr, this.element);
-        if (!isArray(items)) {
-          items = [items];
+        const maxLen = parseInt(this.element.getAttribute('max-items'), 10);
+        if (maxLen < items.length) {
+          items = items.slice(0, maxLen);
         }
-      }
-      user().assert(isArray(items),
-          'Response must contain an array at "%s". %s',
-          itemsExpr, this.element);
-      const maxLen = parseInt(this.element.getAttribute('max-items'), 10);
-      if (maxLen < items.length) {
-        items = items.slice(0, maxLen);
-      }
-      return this.scheduleRender_(items);
-    }, error => {
-      throw user().createError('Error fetching amp-list', error);
-    }).then(() => {
-      if (this.getFallback()) {
-        // Hide in case fallback was displayed for a previous fetch.
-        this.toggleFallbackInMutate_(false);
-      }
-      this.togglePlaceholder(false);
-      this.toggleLoading(false);
+        return this.scheduleRender_(items);
+      }, error => {
+        throw user().createError('Error fetching amp-list', error);
+      }).then(onFetchSuccess_.bind(this),
+          onFetchError_.bind(this));
+    }
+  }
+
+  /**
+   * Proxies the template rendering to the viewer.
+   */
+  fetchAndRenderTemplate_() {
```

Comment:
This name sounds like it's not delegating anything to the viewer. How about `ssrTemplate_()`?